File: data/two-pointers/container-with-most-water.py
def maxArea(height):
    """
    Finds the maximum water that can be stored using two lines in the height array.

    :param height: List[int] - Heights of vertical lines.
    :return: int - Maximum water container area.
    """
    left, right = 0, len(height) - 1  # Two-pointer approach
    max_area = 0  

    while left < right:
        # Calculate area based on current left and right lines
        area = min(height[left], height[right]) * (right - left)
        max_area = max(max_area, area)  # Update max area if needed

        # Move the pointer pointing to the shorter line to try to find a taller one
        if height[left] < height[right]:
            left += 1
        else:
            right -= 1

    return max_area

# Example usage:
height = [1, 8, 6, 2, 5, 4, 8, 3, 7]
print(maxArea(height))  # Output: 49

File: data/two-pointers/container-with-most-water.cpp
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

int maxArea(vector<int>& height) {
    int left = 0, right = height.size() - 1; // Two pointers
    int max_area = 0;

    while (left < right) {
        // Compute area between left and right lines
        int area = min(height[left], height[right]) * (right - left);
        max_area = max(max_area, area);  // Update max area if necessary

        // Move the pointer pointing to the shorter line
        if (height[left] < height[right]) {
            left++;
        } else {
            right--;
        }
    }

    return max_area;
}

// Example usage
int main() {
    vector<int> height = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    cout << "Maximum water that can be contained: " << maxArea(height) << endl;
    return 0;
}
